### PR TITLE
libp2p Requirements Doc - Roadmap to 1.0.0

### PR DESCRIPTION
**Pull request by ghost**
Moved the libp2p items out of https://github.com/mgoelzer/ipfs/blob/master/ROADMAP-TO-1.0.0.md and into this file that will live in libp2p/libp2p

(Note:  after doing a bit of research, there really isn't any good way to do a table with checkboxes inside the cells in Github flavored markdown.  This is a [known issue](https://github.com/github/markup/issues/364) in Github.  The best workaround puts extra vertical whitespace in every cell.  So, instead of using the three column approach, I just triplicated each item for Go, JS and Rust.)

@jbenet @diasdavid 

(Note:  once this PR is merged, https://github.com/ipfs/ipfs/pull/334 should be merged to remove the content from the ipfs/ipfs/ROADMAP-TO-1.0.0.md document.)

### CODE REVIEW

**Comment by daviddias** [comment]
> there really isn't any good way to do a table with checkboxes inside the cells in Github flavored markdown.

True. You have to get creative with emojis, see for example:

- https://github.com/ipfs/js-ipfs/pull/485
- https://github.com/ipfs/ipfs/blob/master/IMPLEMENTATION_STATUS.md

File: ROADMAP-TO-FEATURE-COMPLETION.md [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+### libp2p Modules Implementations 
+
+The libp2p protocols are the protocol stack for the modular libp2p protocols library. These form a p2p-first network stack, with no assumptions, self description, and modular interfaces. More at https://github.com/ipfs/specs/tree/master/libp2p
+
+> Legend: :green_apple: Done &nbsp; :lemon: In Progress &nbsp; :tomato: Missing &nbsp; :chestnut: Not planned
+
+| Transports                                   | Go            | JS - Node.js  |  JS - Browser | Rust          |
+| -------------------------------------------- | :-----------: | :-----------: | :-----------: | :-----------: |
+| **`TCP`**                                    | :green_apple: | :green_apple: | :chestnut:    | :green_apple: |
+| **`WebRTC`**                                 | :tomato:      | :green_apple: | :green_apple: | :tomato:      |
```
Comment (daviddias) [inline ROADMAP-TO-FEATURE-COMPLETION.md]: @mgoelzer here is an example of a table you can use for this section. You can use this to finalize the section and give you that table view you were looking for.

**Comment by daviddias** [comment]
@mgoelzer you missed porting:

- https://github.com/ipfs/ipfs/blob/master/ROADMAP-TO-1.0.0.md#ipfs-dht
- https://github.com/ipfs/ipfs/blob/master/ROADMAP-TO-1.0.0.md#ipfs-pubsub

**Comment by daviddias** [comment]
@mgoelzer any ETA to get this doc finished? I mean, the first pass with the table and all the content items that belong to libp2p so that we can iterate over it.

**Comment by ghost** [comment]
@diasdavid What do you think about merging in the current state so we can iterate?  I fixed up all the fruits and table structure.  [The initial description section](https://github.com/mgoelzer/libp2p/blob/add-roadmap-checklist/REQUIREMENTS.md#description) lists some known deficiencies in the current doc.

Alternative is to keep iterating in this PR.  Either works for me.

Also, these items are intentionally missing:
 * https://github.com/ipfs/ipfs/blob/master/ROADMAP-TO-1.0.0.md#ipfs-dht
 * https://github.com/ipfs/ipfs/blob/master/ROADMAP-TO-1.0.0.md#ipfs-pubsub

They seem like IPFS milestones, and I'm trying to segregate IPFS stuff from pure libp2p stuff.  I left those in the IPFS requirements doc.

**Comment by daviddias** [comment]
image

File: REQUIREMENTS.md [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+ * Rust is incomplete (some tomatoes should be apples)
+ * Should we divide Node.js and Browser, or just a have a single JS column?
+ * Needs a proofread from implementors to make sure the fruits are correct
+ * [Performance Milestones](#performance-milestones) is incomplete - need to agree on what these are
+
+As you improve this document, please remove items from the list above.
+
+### libp2p Modules Implementations 
+
+The libp2p protocols are the protocol stack for the modular libp2p protocols library. These form a p2p-first network stack, with no assumptions, self description, and modular interfaces. More at https://github.com/ipfs/specs/tree/master/libp2p
```
Comment (raulk) [inline REQUIREMENTS.md]: Link broken. New one: https://github.com/libp2p/specs/

**Comment by raulk** [comment]
@mgoelzer @diasdavid a few thoughts:

1. By _Node_, do we mean _Host_ in libp2p terminology?
2. Personally I consider the enumerated technologies/protocols listed as concrete _implementations_ rather than _interfaces_. 
3. Do we have criteria or a definition for _done_ for each module? If not, I wonder if the term _available_ would be more appropriate, as it conveys that an implementation exists, but does not set any expectations on feature completeness. To me, _done_ also implies some kind of closure, which may be applicable in the context of specific feature sets we aim for with 1.0 for each module.

**Comment by daviddias** [comment]
> By Node, do we mean Host in libp2p terminology?

It is actually the other way. Host is a legacy term, the correct term is libp2p Node.

> Personally I consider the enumerated technologies/protocols listed as concrete implementations rather than interfaces.

Agreed. Same that I was going with https://github.com/libp2p/libp2p/pull/39#issuecomment-414979622

**Comment by ghost** [comment]
>[screenshot]

@diasdavid Yes, I had the same reaction as you. `impl`(emented) is already implied for every row in the table, unless otherwise stated, and the word `interfaces` doesn't make much sense in the section headings.  I (intentionally) left those alone because I wanted a verbatim copy of the original text in https://github.com/mgoelzer/ipfs/blob/master/ROADMAP-TO-1.0.0.md, thus making this PR a pure move-and-reformat with no change in semantics, which tend to be non-controversial.

That said, if the terms have no meaning to you, then they probably aren't of significance to anyone else.  I'm fine with taking them out.

**Comment by ghost** [comment]
@raulk 

> Do we have criteria or a definition for done

I suppose nothing is ever truly "done" in software, but I think of it as meaning we have an implementation that is suitable for downstreams to use in production software.  If these modules were independently SemVer2 versioned, we could use that to imply an additional statement about our commitment not to break backwards compat, but the libp2p project is not at that stage of maturity yet.  There's a lot of fundamental innovation that still needs to happen.

> wonder if the term available would be more appropriate

I totally agree with the principle of what you're saying, and I'm very tempted to make the change to `Available`, but the `Done`/`InProgress`/`Missing`/`NotPlanned` emoji convention is being used [elsewhere](https://github.com/ipfs/ipfs/blob/master/IMPLEMENTATION_STATUS.md) in PL projects.  I'm somewhat reluctant to redefine one term here. I don't have super strong feelings though.

>By Node, do we mean Host in libp2p terminology?

Yes, they are the same.  This is a long standing problem, and unfortunately it's not the only place where you'll find language implementation divergence as you dig in deeper.  Juan has been suggesting for a while that we adopt a standardized interface (e.g., [see here](https://gist.github.com/jbenet/c8e8d85fcd5a06b3fb75b3690ba02aaf)), specified in a golang-like DSL, but it's harder than it sounds because different languages have fundamentally different paradigms and idioms.  An example would be channels in Go vs callbacks in JS vs Sync/Send traits and std::thread in Rust, all of which achieve a similar goal (concurrency, or the appearance of it in single-threaded runtimes) but are more than just syntactically different; there's an actual conceptual difference. So, this turns out to be a somewhat hard coordination problem.

File: REQUIREMENTS.md (hunk 2 above)
Comment (ghost) [reply REQUIREMENTS.md]: Thanks - good catch

**Comment by ghost** [comment]
I pushed a new commit to remove "interface", "impl" and "implementation" per https://github.com/libp2p/libp2p/pull/39#issuecomment-414979622.

Did not change "Node" to "Host" per https://github.com/libp2p/libp2p/pull/39#issuecomment-415019885

Did not change "Done" to "Available" based on consistency-with-IPFS argument in https://github.com/libp2p/libp2p/pull/39#issuecomment-415110288 (I'm open to changing this later though)

I'd like to merge this, merge https://github.com/ipfs/ipfs/pull/334, and then we can iterate on this one and let IPFS iterate on its analogous list.  In subsequent PRs, I'd like to (1) verify the accuracy of the fruits, (2) fill in `Performance Milestones` TODO item with 10^6 and 10^9 node goals, and (2) add some docs requirements line items.

**Comment by ghost** [comment]
@diasdavid Forgot to respond to this:

>@mgoelzer you missed porting:
>https://github.com/ipfs/ipfs/blob/master/ROADMAP-TO-1.0.0.md#ipfs-dht
>https://github.com/ipfs/ipfs/blob/master/ROADMAP-TO-1.0.0.md#ipfs-pubsub

I have [another branch](https://github.com/mgoelzer/libp2p/blob/add-requirements-dht-pubsub/REQUIREMENTS.md) that adds on to this PR where I have included those.  My next PR will be based off of that, with some additional stuff mentioned above.

**Pull request merged**